netbox-prometheus-sd.py

```diff
@@ -6,88 +6,180 @@
-            if getattr(device, 'parent_device', None):
-                labels['__meta_netbox_parent'] = device.parent_device.name
-            if getattr(device, 'address', None):
-                labels['__meta_netbox_address'] = device.address
-            if getattr(device, 'description', None):
-                labels['__meta_netbox_description'] = device.description
-            try:
-                device_targets = json.loads(device.custom_fields[args.custom_field])
-            except ValueError:
-                continue  # Ignore errors while decoding the target json FIXME: logging
-
-            if not isinstance(device_targets, list):
-                device_targets = [device_targets]
-
-            for target in device_targets:
-                target_labels = labels.copy()
-                target_labels.update(target)
-                if hasattr(device, 'primary_ip'):
-                    address = device.primary_ip
+class Discovery(object):
+    def __init__(self, args):
+        super(Discovery, self).__init__()
+        self.args = args
+
+    def run(self):
+        self.netbox = pynetbox.api(self.args.url, token=self.args.token)
+        if self.args.discovery == 'device':
+            targets = self.discover_device()
+        elif self.args.discovery == 'circuit':
+            targets = self.discover_circuit()
+        else:
+            return
+
+        temp_file = None
+        if self.args.output == '-':
+            output = sys.stdout
+        else:
+            temp_file = '{}.tmp'.format(self.args.output)
+            output = open(temp_file, 'w')
+
+        json.dump(targets, output, indent=4)
+        output.write('\n')
+
+        if temp_file:
+            output.close()
+            os.rename(temp_file, self.args.output)
+        else:
+            output.flush()
+
+    def discover_device(self):
+        targets = []
+
+        # Filter out devices without primary IP address as it is a requirement
+        # to be polled by Prometheus
+        devices = self.netbox.dcim.devices.filter(has_primary_ip=True)
+        vm = self.netbox.virtualization.virtual_machines.filter(
+            has_primary_ip=True)
+        ips = self.netbox.ipam.ip_addresses.filter(
+            **{'cf_%s' % self.args.custom_field: '{'})
+
+        for device in itertools.chain(devices, vm, ips):
+            if device.custom_fields.get(self.args.custom_field):
+                labels = {'__port__': str(self.args.port)}
+                if getattr(device, 'name', None):
+                    labels['__meta_netbox_name'] = device.name
                 else:
-                    address = device
-                targets.append({'targets': ['%s:%s' % (str(netaddr.IPNetwork(address.address).ip),
-                                                       target_labels['__port__'])],
-                                'labels': target_labels})
+                    labels['__meta_netbox_name'] = repr(device)
```

Comment:
Same comment as the previous one, you are breaking compatibility for existing users. If you think this change can be useful, motivate it and isolate it in a separated commit.